File: searchInfiniteSortedArr.cpp
// Time Complexity : O(log m * log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int search(const ArrayReader& reader, int target) {

        int low = 0;
        int high = 1;

        // This takes time complexity of O(log m)
        while(reader.get(high) < target) {
            low = high;
            high *= 2;
        }
        // This takes time complexity of O(log n), where n ranges between low and high
        while(low <= high) {
            int mid = low + (high - low)/2;
            if(reader.get(mid) == target) {
                return mid;
            } else if(reader.get(mid) > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return -1;
    }
};

File: SearchInfiniteSortedArr.py
# // Time Complexity : log(m) + log(n), where m = range of values, n is binary search
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : No it's a leetcode premium problem
# // Any problem you faced while coding this : None

class Solution:
    def search(self, reader: 'ArrayReader', target: int) -> int:

        if (reader.get(0) == target): 
            return 0
        low = 0
        high = 1
      #get the range of the array
        while reader.get(high) < target:
            low = high
            high = 2 * high
      #Regular binary search
        while low <= high:
            mid = low + (high - low) // 2

            if reader.get(mid) == target:
                return mid
            elif target < reader.get(mid):
                high = mid - 1
            elif target > reader.get(mid):
                low = mid + 1
        return -1